File: atcoder/tyama_atcoderxmascontest2015A.py
#!/usr/bin/python
n,x,t,a,b,c=map(int,open(0).read().split());g=pow(a,t,c);h=pow(a,t,c*~-a)//~-a*b;s=0
for _ in[0]*n:s+=x;x=(x*g+h)%c
print(s)
#z=lambda n,x:n and x+z(n-1,(x*g+h)%c);print(z(n,x))

File: atcoder/tyama_atcoderxmascontest2015A.rb
#!/usr/bin/ruby
require'openssl';n,x,t,a,b,c=`dd`.split.map &:to_i
g=a.to_bn.mod_exp(t,c)
h=a.to_bn.mod_exp(t,c*a-=1)/a
s=0
n.times{s+=x;x=(x*g+b*h[0])%c}
p s
